src/Microsoft.Identity.Web/AzureADB2COpenIDConnectEventHandlers.cs

```diff
@@ -10,103 +10,103 @@
-            }
-
-            // Access denied errors happen when a user cancels an action on the Azure Active Directory B2C UI. We just redirect back to
-            // the main page in that case.
-            // Message contains error: 'access_denied', error_description: 'AADB2C90091: The user has canceled entering self-asserted information.
-            // Correlation ID: d01c8878-0732-4eb2-beb8-da82a57432e0
-            // Timestamp: 2018-03-05 02:56:49Z
-            // ', error_uri: 'error_uri is null'.
-            else if (context.Failure is OpenIdConnectProtocolException && context.Failure.Message.Contains(ErrorCodes.AccessDenied))
-            {
-                context.Response.Redirect($"{context.Request.PathBase}/");
-            }
-            else
-            {
-                _errorAccessor.SetMessage(context.HttpContext, context.Failure?.Message);
-
-                context.Response.Redirect($"{context.Request.PathBase}/MicrosoftIdentity/Account/Error");
-            }
-
-            return Task.CompletedTask;
-        }
-
-        private string BuildIssuerAddress(RedirectContext context, string? defaultUserFlow, string userFlow)
-        {
-            if (!_userFlowToIssuerAddress.TryGetValue(userFlow, out var issuerAddress))
-            {
-                _userFlowToIssuerAddress[userFlow] = context.ProtocolMessage.IssuerAddress.ToLowerInvariant()
-                    .Replace($"/{defaultUserFlow?.ToLowerInvariant()}/", $"/{userFlow.ToLowerInvariant()}/");
-            }
-
-            return _userFlowToIssuerAddress[userFlow];
-        }
-    }
+	internal class AzureADB2COpenIDConnectEventHandlers
+	{
+		private readonly ILoginErrorAccessor _errorAccessor;
+
+		private readonly IDictionary<string, string> _userFlowToIssuerAddress =
+			new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+		public AzureADB2COpenIDConnectEventHandlers(
+			string schemeName,
+			MicrosoftIdentityOptions options,
+			ILoginErrorAccessor errorAccessor)
+		{
+			SchemeName = schemeName;
+			Options = options;
+			_errorAccessor = errorAccessor;
+		}
+
+		public string SchemeName { get; }
+
+		public MicrosoftIdentityOptions Options { get; }
+
+		public Task OnRedirectToIdentityProvider(RedirectContext context)
+		{
+			var defaultUserFlow = Options.DefaultUserFlow;
+			if (context.Properties.Items.TryGetValue(OidcConstants.PolicyKey, out var userFlow) &&
+				!string.IsNullOrEmpty(userFlow) &&
+				!string.Equals(userFlow, defaultUserFlow, StringComparison.OrdinalIgnoreCase))
+			{
+				context.ProtocolMessage.IssuerAddress = BuildIssuerAddress(context, defaultUserFlow, userFlow);
+				context.Properties.Items.Remove(OidcConstants.PolicyKey);
+
+				if (!Options.HasClientCredentials)
+				{
+					context.ProtocolMessage.ResponseType = OpenIdConnectResponseType.IdToken;
+				}
+				else
+				{
+					if (Options.IsB2C)
```

Comment:
This if block can be less nested by combining towards "else if" which will slightly improve readability.